File: glpython/glpython/background.py
#!/usr/bin/env python

#------------------------------------------------------------------------------
#
#   Copyright (c) 2007 Nicolas Rougier
# 
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
# 
#------------------------------------------------------------------------------
""" Background object

A background is a simple filled rectangular area (relative to window) area
with a border.
"""

import OpenGL.GL as GL
from object import Object


class Background (Object):
    """ Background object """
    
    def __init__(self):
        """ Intialization with default values"""
        
        Object.__init__(self)
        self.bg_color_bottom = (1, 1, 1, 0)
        self.bg_color_top = (1, 1, 1, 0)
        self.br_color = (0,0,0,1)
        self.use_border = False

    
    def render(self):
        """ Render the background with proper "2D" view """
        
        viewport = GL.glGetIntegerv (GL.GL_VIEWPORT)
        w,h = viewport[2], viewport[3]

        # Switch to '2D' mode
        GL.glPushAttrib (GL.GL_ENABLE_BIT)        
        GL.glMatrixMode (GL.GL_PROJECTION)
        GL.glPushMatrix ()
        GL.glLoadIdentity ()
        GL.glOrtho (0, w, 0, h, -100, 100)
        GL.glMatrixMode (GL.GL_MODELVIEW)
        GL.glPushMatrix ()
        GL.glLoadIdentity()

        # Disable what is not necessary
        GL.glDisable (GL.GL_LIGHTING)
        GL.glDisable (GL.GL_TEXTURE_2D)
        GL.glEnable (GL.GL_BLEND)
        GL.glBlendFunc (GL.GL_SRC_ALPHA, GL.GL_ONE_MINUS_SRC_ALPHA)
        GL.glDisable (GL.GL_DEPTH_TEST)

        # Background
        GL.glTranslatef (0.0, 0.0, -99)
        GL.glPolygonMode (GL.GL_FRONT_AND_BACK, GL.GL_FILL)
        GL.glBegin (GL.GL_QUADS)
        GL.glColor (self.bg_color_bottom)
        GL.glVertex2i (0, 0)
        GL.glVertex2i (w-1, 0)
        GL.glColor (self.bg_color_top)
        GL.glVertex2i (w-1, h-1)
        GL.glVertex2i (0, h-1)
        GL.glEnd()

        # Border
        if self.use_border:
            GL.glTranslatef (0.0, 0.0, 1)        
            GL.glPolygonMode (GL.GL_FRONT_AND_BACK, GL.GL_LINE)
            GL.glBegin (GL.GL_QUADS)
            GL.glColor (self.br_color)
            GL.glVertex2i (0, 0)
            GL.glVertex2i (w-1, 0)
            GL.glVertex2i (w-1, h-1)
            GL.glVertex2i (0, h-1)
            GL.glEnd()        
 
        # Go back to previous projection and view
        GL.glMatrixMode (GL.GL_PROJECTION)
        GL.glPopMatrix ()
        GL.glMatrixMode (GL.GL_MODELVIEW)
        GL.glPopMatrix()
        GL.glPopAttrib()

File: glpython/glpython/objects/background.cc
//
// Copyright (C) 2007 Nicolas Rougier
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as
// published by the Free Software Foundation; either version 2 of the
// License, or (at your option) any later version.
//
// $Id$

#include "../core/colormaps.h"
#include "background.h"

using namespace glpython;
using namespace glpython::objects;


//___________________________________________________________________Background
Background::Background (core::ColormapPtr cmap,
                        std::string orientation,
                        float alpha, 
                        std::string name) : core::Object (name)
{
    set_cmap (cmap);
    set_alpha (alpha);
    set_orientation (orientation);
    depth = 0;
    list = 0;
}

//__________________________________________________________________~Background
Background::~Background (void)
{
    if (list)
        glDeleteLists (list,1);
}

//_________________________________________________________________get/set cmap
void
Background::set_cmap (core::ColormapPtr cmap)
{
    if (cmap->colors.size() < 2) {
        PyErr_SetString (PyExc_ValueError, 
                         "Colormap must have at least 2 values.");
        return;
    }

    this->cmap = cmap;
    this->dirty = true;   
}

core::ColormapPtr
Background::get_cmap (void)
{
    return cmap;
}

//__________________________________________________________get/set orientation
void
Background::set_orientation (std::string orientation)
{
    if (orientation == "horizontal")
        this->orientation = "horizontal";
    else
        this->orientation = "vertical";
    this->dirty = true;
}

std::string
Background::get_orientation (void)
{
    return orientation;
}

//________________________________________________________________get/set alpha
void
Background::set_alpha (float alpha)
{
    this->alpha = alpha;
    if (this->alpha > 1.0f)
        this->alpha = 1.0f;
    else if (this->alpha < 0.0f)
        this->alpha = 0.0f;
    this->dirty = true;    
}
float
Background::get_alpha (void)
{
    return alpha;
}

//_______________________________________________________________________render
void
Background::render (void)
{
    if (cmap->colors.size() < 2)
        return;

    GLint mode;
    glGetIntegerv (GL_RENDER_MODE, &mode);
    
    if (mode == GL_SELECT)
        return;
    if (!visible)
        return;
    if (!dirty) {
        glCallList (list);
        return;
    }


    int viewport[4];
    glGetIntegerv (GL_VIEWPORT, viewport);
    int x = 0;
    int y = 0;
    int w = viewport[2];
    int h = viewport[3];

    if (list)
        glDeleteLists (list,1);
    list = glGenLists(1);

    glNewList (list, GL_COMPILE_AND_EXECUTE);
    glPolygonMode (GL_FRONT_AND_BACK, GL_FILL);
    glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
    glEnable(GL_BLEND);
    glMatrixMode (GL_PROJECTION);
    glPushMatrix ();
    glLoadIdentity ();
    glOrtho (0, w, 0, h, -1, 1);
    glMatrixMode (GL_MODELVIEW);
    glPushMatrix ();
    glLoadIdentity();
    glDisable (GL_DEPTH_TEST);

    // Colors
    for (unsigned int i=0; i<(cmap->colors.size()-1); i++) {
        int x1 = 0, y1 = 0, x2 = 0, y2 = 0;

        glBegin (GL_QUADS);        
        if (orientation == "vertical") {
            x1 = x;
            x2 = x+w;
            y1 = y+int(h*cmap->colors[i].data[core::VALUE]);
            y2 = y+int(h*cmap->colors[i+1].data[core::VALUE]);
            glColor4f (cmap->colors[i].data[0], cmap->colors[i].data[1],
                       cmap->colors[i].data[2], cmap->colors[i].data[3]*alpha);
            glVertex2i (x1, y1);
            glVertex2i (x2, y1);
            glColor4f (cmap->colors[i+1].data[0], cmap->colors[i+1].data[1],
                       cmap->colors[i+1].data[2], cmap->colors[i+1].data[3]*alpha);
            glVertex2i (x2, y2);
            glVertex2i (x1, y2);
        } else {
            x1 = x+int(w*cmap->colors[i].data[core::VALUE]);
            x2 = x+int(w*cmap->colors[i+1].data[core::VALUE]);
            y1 = y;
            y2 = y+h;
            glColor4f (cmap->colors[i].data[0], cmap->colors[i].data[1],
                       cmap->colors[i].data[2], cmap->colors[i].data[3]*alpha);
            glVertex2i (x1, y1);
            glVertex2i (x1, y2);
            glColor4f (cmap->colors[i+1].data[0], cmap->colors[i+1].data[1],
                       cmap->colors[i+1].data[2], cmap->colors[i+1].data[3]*alpha);
            glVertex2i (x2, y2);
            glVertex2i (x2, y1);
        }
        glEnd();
    }

    glEnable (GL_DEPTH_TEST);
    glMatrixMode (GL_PROJECTION);
    glPopMatrix ();
    glMatrixMode (GL_MODELVIEW);
    glPopMatrix();    
    glEndList();
    dirty = false;
}

//________________________________________________________________python_export
void
Background::python_export (void) {

    using namespace boost::python;
    register_ptr_to_python< boost::shared_ptr<Background> >();

    class_<Background, bases <core::Object> >("Background",
    " ______________________________________________________________________\n"
    "                                                                       \n"
    " ______________________________________________________________________\n",
     
    init< core::ColormapPtr, optional <std::string, float, std::string > > (
        (arg("cmap") = core::Colormaps::Default,
         arg("orientation") = "vertical",
         arg("alpha") = 1.0,
         arg("name") = "Background"),
        "__init__ (cmap, orientation, alpha, name)\n"))

    .add_property ("cmap",         &Background::get_cmap,        &Background::set_cmap)
    .add_property ("orientation",  &Background::get_orientation, &Background::set_orientation)
    .add_property ("alpha",        &Background::get_alpha,       &Background::set_alpha)
    ;       
}
